lib/services/local/authinfo.go

```diff
@@ -0,0 +1,111 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package local
+
+import (
+	"context"
+
+	"github.com/coreos/go-semver/semver"
+	"github.com/gravitational/trace"
+
+	"github.com/gravitational/teleport/api/gen/proto/go/teleport/authinfo/v1"
+	"github.com/gravitational/teleport/api/types"
+	authinfotype "github.com/gravitational/teleport/api/types/authinfo"
+	"github.com/gravitational/teleport/lib/backend"
+	"github.com/gravitational/teleport/lib/services"
+)
+
+const (
+	// authInfoPrefix is a backend key for storing the auth server information.
+	authInfoPrefix = "auth_info"
+)
+
+// AuthInfoService is responsible for managing the information about auth server.
+type AuthInfoService struct {
+	backed backend.Backend
+}
+
+// NewAuthInfoService returns a new AuthInfoService.
+func NewAuthInfoService(b backend.Backend) (*AuthInfoService, error) {
+	return &AuthInfoService{
+		backed: b,
+	}, nil
+}
+
+// GetTeleportVersion reads the last known Teleport version from storage.
+func (s *AuthInfoService) GetTeleportVersion(ctx context.Context) (*semver.Version, error) {
+	item, err := s.backed.Get(ctx, backend.NewKey(authInfoPrefix))
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	info, err := services.UnmarshalProtoResource[*authinfo.AuthInfo](item.Value)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if err := authinfotype.ValidateAuthInfo(info); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	version, err := semver.NewVersion(info.GetSpec().TeleportVersion)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return version, nil
+}
+
+// WriteTeleportVersion writes the last known Teleport version to the storage.
+func (s *AuthInfoService) WriteTeleportVersion(ctx context.Context, version *semver.Version) error {
+	if version == nil {
+		return trace.BadParameter("wrong version parameter")
+	}
+
+	info, err := authinfotype.NewAuthInfo(&authinfo.AuthInfoSpec{
+		TeleportVersion: version.String(),
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if err := authinfotype.ValidateAuthInfo(info); err != nil {
+		return trace.Wrap(err)
+	}
+	rev, err := types.GetRevision(info)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	value, err := services.MarshalProtoResource[*authinfo.AuthInfo](info)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	expires, err := types.GetExpiry(info)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	item := backend.Item{
+		Key:      backend.NewKey(authInfoPrefix),
+		Value:    value,
+		Expires:  expires,
+		Revision: rev,
+	}
+	if _, err := s.backed.Put(ctx, item); err != nil {
+		return trace.Wrap(err)
+	}
```

Comment:
The unconditional put makes racing possible, someone might have updated the version in the meantime. This operation should use conditional updates based on revision. I think we have all the machinery available in the generic service struct.

Reply:
I'm Changing this one, but not sure if it is good to fail auth init because when we do parallel scale of auth instances and spawning same auth versions in cluster, they have to conditionally write to the same key and theoretically if they launched and the same second one of the instances must fail to write and restart